### PR TITLE
[jit] Support properties on `Device`

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #32955 - [jit] Fix flipped PackedSequence outputs in script
 * **#32953 - [jit] Support properties on `Device`**

PyTorch devices have a `index` and `type` property. This PR adds support for both to TorchScript

Differential Revision: [D19849320](https://our.internmc.facebook.com/intern/diff/19849320/)

### CODE REVIEW

File: torch/csrc/jit/script/sugared_value.cpp [hunk 1]
```diff
@@ -89,6 +89,19 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
     }
   }
 
+  if (value_->type()->isSubtypeOf(DeviceObjType::get())) {
+    // if (value_->type()->kind() == TypeKind::DeviceObjType) {
+    // property lookups on Device
+    static const std::unordered_set<std::string> fields = {
```
Comment (eellison) [inline torch/csrc/jit/script/sugared_value.cpp]: Can you refactor this with the tensor fields above ? Maybe something like:
`static const std::unordered_map<TypeKind, std::unordered_set<std::string>> builtin_type_properties`

**Review by eellison — APPROVED** [review]
Could you refactor this with the tensor properties ?

**Comment by eellison** [comment]
If you want to tack on https://github.com/pytorch/pytorch/issues/26731 to this PR be my quest  😃

File: torch/csrc/jit/script/sugared_value.cpp [hunk 2]
```diff
@@ -67,22 +67,29 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
...
+                                 "data",
+                                 "shape",
+                                 "is_cuda",
+                                 "is_sparse",
+                                 "is_mkldnn",
+                                 "is_quantized",
+                                 "requires_grad",
+                                 "layout",
+                             }},
+                            {TypeKind::DeviceObjType, {"type", "index"}}};
```
Comment (eellison) [inline torch/csrc/jit/script/sugared_value.cpp]: why is there "index" on device type here ?

Comment (driazati) [reply torch/csrc/jit/script/sugared_value.cpp]: forgot to implement it, it's added in the latest commit

**Comment by driazati** [comment]
#26731 is a longer way off since it'd require a whole new type and `IValue` since data types aren't first class like `Device` is

**Pull request closed without merging**